### PR TITLE
[7.x] Allow Broadcast Authorization Callbacks To Accept "Guests"

### PR DESCRIPTION
**Pull request by matheusb-comp**
This PR allows Broadcast channels [authorization callbacks](https://laravel.com/docs/6.x/broadcasting#defining-authorization-callbacks) to define "guest users", based on how [`Gates` do](https://laravel.com/docs/6.x/authorization#guest-users), since #24576.

This implements Idea [#1227](https://github.com/laravel/ideas/issues/1227), with the checking methods pretty much copied from the ones currently used by [`Gate`](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Auth/Access/Gate.php#L378).

At the moment, both the `PusherBroadcaster` and `RedisBroadcaster` throw `AccessDeniedHttpException` when [`retrieveUser()`](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Broadcasting/Broadcasters/PusherBroadcaster.php#L46) returns `null`, during private channels authorizations, before executing the user-defined authorization callback:
```
if ($this->isGuardedChannel($request->channel_name) &&
    ! $this->retrieveUser($request, $channelName)) {
    throw new AccessDeniedHttpException;
}
```

Therefore, it is impossible to use private channels with authentication methods that are not tied to a `User`, like Passport's [Client Credentials Grant](https://laravel.com/docs/6.x/passport#client-credentials-grant-tokens), that is used for machine-to-machine communication, as pointed out in Passport issue [#1166](https://github.com/laravel/passport/issues/1166).

Also, I imagine that the current check is in place to prevent the `$user` variable from being `null` in the authorization callback. However, if this code was executed when subscribing to a public channel, the `isGuardedChannel()` method would return `false`, and the `$user` variable could be passed with a `null` value to the authorization callback. I believe this doesn't happen because no HTTP request is made when subscribing to a public channel.

But with the methods from the `Gate` guest user authorization, this depends on how the callback is defined, and a `null` `$user` will only be allowed when an `optional` type-hint or a `null` default value is defined:
```
protected function parameterAllowsGuests($parameter)
{
    return ($parameter->getClass() && $parameter->allowsNull()) ||
           ($parameter->isDefaultValueAvailable() && is_null($parameter->getDefaultValue()));
}
```

### CODE REVIEW

File: src/Illuminate/Broadcasting/Broadcasters/Broadcaster.php [hunk 1]
```diff
@@ -73,14 +73,63 @@ protected function verifyUserCanAccessChannel($request, $channel)
...
         }
 
         throw new AccessDeniedHttpException;
     }
 
+    /**
+     * Determine whether the callback can be called with the given user.
+     *
+     * @param  \Illuminate\Contracts\Auth\Authenticatable|null  $user
+     * @param  \Closure|callable $callback
```
Comment (lucasmichot) [inline src/Illuminate/Broadcasting/Broadcasters/Broadcaster.php]: `\Closure` is already callable, so `@param  callable  $callback` is enough

Comment (matheusb-comp) [reply src/Illuminate/Broadcasting/Broadcasters/Broadcaster.php]: Thanks!

**Comment by taylorotwell** [comment]
So, can you walk me through your specific application scenario that led you to want this?

**Comment by matheusb-comp** [comment]
> So, can you walk me through your specific application scenario that led you to want this?

Laravel has to communicate with a small group of "workers" written in JavaScript running on a different server, and the status of the workers (online / loading / ready / offline) can change at any time and must be always known by the Laravel application.

Initially I thought about using a pub-sub server where the workers could subscribe to a "data" channel and send their status through a "control" channel that the Laravel application could subscribe to. However the application would need to keep track of the workers, and there would need to be some sort of authentication with the pub-sub server, essentially "private channels".

Then a solution with the broadcasting system and Laravel Echo seemed easier, it already has the connect/disconnet and ping/pong, private and presence channels, and I can run the WebSocket server in the same place as the Laravel application (to receive the worker's status).
Also, both broadcasting (to browser clients) and Passport are already in the application.

Since the workers are not related to any user, they are authenticated using the client credentials grant. When setting up the private channel authorization and it not working, I tried authorization in normal routes (Gate) to make sure the Passport grant was correct, and found the [guest user docs](https://laravel.com/docs/6.x/authorization#guest-users).

**Comment by matheusb-comp** [comment]
After more consideration, in my specific scenario this is too much. I can just make the pub-sub connection port available only for the "workers", and then they don't need any authentication. This also removes the need to run the WebSocket server locally (then a service like Pusher could be used).

However, apparently other people also tried to use the broadcasting system with the client credentials grant.
And unlike the discussion in #24576, this is not a case of authorization without authentication, the "user" in this case was authenticated by a valid oauth token, it is just not associated with the `User` model.

Since the [Events](https://laravel.com/docs/7.x/events) are not required to be associated with a `User` and they can implement the `shouldBroadcast` interface, a "system event" could be created not intended for the "human users" (or machines holding oauth tokens that represent them).

It is then strange that machines with oauth client credentials that don't represent human users can be authenticated and even be authorized by gates, but can't be authorized to receive broadcasts in private channels.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by chimit** [comment]
It would be very useful if we can fully control broadcasting auth in the `channels.php`. I believe it would be more consistent if broadcasting auth works in a similar way as Policies. Currently, we have to use some [workarounds](https://stackoverflow.com/questions/43341820/laravel-echo-allow-guests-to-connect-to-presence-channel#).

In my scenario, I have a chat app that allows guests to join some channels. I can't use public channels because I still need to know who from authenticated users is connected (losing `.here`, `.joining`, `.leaving` events in Echo).

**Pull request closed without merging**